src/mapping/query/arithmetic.rs

```diff
@@ -0,0 +1,406 @@
+use super::Function;
+use crate::{
+    event::{Event, Value},
+    mapping::Result,
+};
+use bytes::BytesMut;
+
+#[derive(Debug, Clone)]
+pub(in crate::mapping) enum Operator {
+    Multiply,
+    Divide,
+    Modulo,
+    Add,
+    Subtract,
+    Equal,
+    NotEqual,
+    Greater,
+    GreaterOrEqual,
+    Less,
+    LessOrEqual,
+    And,
+    Or,
+}
+
+#[derive(Debug)]
+pub(in crate::mapping) struct Arithmetic {
+    left: Box<dyn Function>,
+    right: Box<dyn Function>,
+    op: Operator,
+}
+
+impl Arithmetic {
+    pub(in crate::mapping) fn new(
+        left: Box<dyn Function>,
+        right: Box<dyn Function>,
+        op: Operator,
+    ) -> Self {
+        Self { left, right, op }
+    }
+}
+
+/// If either value is a floating point number type and the other value is an
+/// integer type the integer is "degraded" into a float. This allows us to
+/// perform arithmetic on common values, but if both are integers then their
+/// precision is preserved.
+fn consistent_number_types(mut left: Value, mut right: Value) -> (Value, Value) {
```

Comment:
I'd suggest "coerce" instead of "consistent", as that is the terminology I have seen most commonly for this operation.